### PR TITLE
HBASE-23117: Bad enum in hbase:meta info:state column can fail loadMeta and stop startup

### PR DESCRIPTION
**Pull request by sandeepvinayak**

### CODE REVIEW

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/RegionStateStore.java [hunk 1]
```diff
@@ -352,8 +352,16 @@ public static State getRegionState(final Result r, int replicaId) {
...
-    return State.valueOf(Bytes.toString(cell.getValueArray(), cell.getValueOffset(),
-        cell.getValueLength()));
+
+    String state = Bytes.toString(cell.getValueArray(), cell.getValueOffset(),
+        cell.getValueLength());
+    try {
+      return State.valueOf(state);
+    }
+    catch (IllegalArgumentException e) {
+      LOG.debug("BAD value {} in hbase:meta info:state column", state);
```
Comment (wchevreuil) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/RegionStateStore.java]: Can we also log the specific region encoded name? Also, as it's an exceptional situation, I think log level should be WARN. We could also give some tip for fixing it using hbck2 _setRegionStateInMeta_ command?

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   6m 54s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -0 :warning: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ branch-2 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   8m 12s |  branch-2 passed  |
| +1 :green_heart: |  compile  |   1m 21s |  branch-2 passed  |
| +1 :green_heart: |  checkstyle  |   1m 45s |  branch-2 passed  |
| +1 :green_heart: |  shadedjars  |   5m 48s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 44s |  branch-2 passed  |
| +0 :ok: |  spotbugs  |   4m 36s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   4m 33s |  branch-2 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   5m 56s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 57s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 57s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 27s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   4m 39s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |  17m 16s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| +1 :green_heart: |  javadoc  |   0m 34s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   3m 36s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 292m 47s |  hbase-server in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 40s |  The patch does not generate ASF License warnings.  |
|  |   | 364m 21s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestSnapshotTemporaryDirectoryWithRegionReplicas |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/867 |
| JIRA Issue | HBASE-23117 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 9159de90a3c4 4.15.0-66-generic #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-867/out/precommit/personality/provided.sh |
| git revision | branch-2 / 12b2e1fd52 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/1/testReport/ |
| Max. process+thread count | 4847 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/RegionStateStore.java (hunk 1 above)
Comment (saintstack) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/RegionStateStore.java]: What @wchevreuil  said

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   2m 13s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ branch-2 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   8m 37s |  branch-2 passed  |
| +1 :green_heart: |  compile  |   1m 13s |  branch-2 passed  |
| +1 :green_heart: |  checkstyle  |   1m 51s |  branch-2 passed  |
| +1 :green_heart: |  shadedjars  |   5m 45s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 49s |  branch-2 passed  |
| +0 :ok: |  spotbugs  |   4m 25s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   4m 22s |  branch-2 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   6m 46s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  1s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  1s |  the patch passed  |
| -1 :x: |  checkstyle  |   1m 35s |  hbase-server: The patch generated 2 new + 223 unchanged - 0 fixed = 225 total (was 223)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   4m 35s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |  16m 59s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| +1 :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   3m 47s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 261m  5s |  hbase-server in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 26s |  The patch does not generate ASF License warnings.  |
|  |   | 328m 40s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/867 |
| JIRA Issue | HBASE-23117 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux b24cd14b8338 4.15.0-66-generic #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-867/out/precommit/personality/provided.sh |
| git revision | branch-2 / c8592f1fb7 |
| Default Java | 1.8.0_181 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/2/artifact/out/diff-checkstyle-hbase-server.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/2/testReport/ |
| Max. process+thread count | 4943 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/assignment/TestRegionStateStore.java [hunk 2]
```diff
@@ -84,6 +89,36 @@ public void visitRegionState(Result result, RegionInfo regionInfo, RegionState.S
...
+    try (Table table = util.getConnection().getTable(TableName.META_TABLE_NAME)) {
+      table.put(put);
+    }
+
+    final AtomicBoolean visitorCalled = new AtomicBoolean(false);
+    regionStateStore.visitMetaForRegion(encodedName, new RegionStateStore.RegionStateVisitor() {
+      @Override
+      public void visitRegionState(Result result, RegionInfo regionInfo, RegionState.State state,
+                                   ServerName regionLocation, ServerName lastHost, long openSeqNum) {
+        assertEquals(encodedName, regionInfo.getEncodedName());
```
Comment (guangxuCheng) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/master/assignment/TestRegionStateStore.java]: Need to check if the state is null

**Comment by guangxuCheng** [comment]
checkstyle warnings should also be fixed.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/assignment/TestRegionStateStore.java (hunk 2 above)
Comment (wchevreuil) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/master/assignment/TestRegionStateStore.java]: Agree with @guangxuCheng comments above.

Comment (sandeepvinayak) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/master/assignment/TestRegionStateStore.java]: @guangxuCheng @wchevreuil  done, thanks for pointing it out.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   2m  0s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ branch-2 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   7m 44s |  branch-2 passed  |
| +1 :green_heart: |  compile  |   1m 18s |  branch-2 passed  |
| +1 :green_heart: |  checkstyle  |   1m 51s |  branch-2 passed  |
| +1 :green_heart: |  shadedjars  |   5m 42s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 46s |  branch-2 passed  |
| +0 :ok: |  spotbugs  |   4m 21s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   4m 18s |  branch-2 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   6m 45s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 13s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 13s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 42s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   5m  5s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |  21m 30s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| +1 :green_heart: |  javadoc  |   0m 50s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   4m 24s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 240m 51s |  hbase-server in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 26s |  The patch does not generate ASF License warnings.  |
|  |   | 315m 42s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/867 |
| JIRA Issue | HBASE-23117 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 0dafe91aed4b 4.15.0-66-generic #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-867/out/precommit/personality/provided.sh |
| git revision | branch-2 / 11d0925814 |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/3/testReport/ |
| Max. process+thread count | 5030 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-867/3/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by sandeepvinayak** [comment]
@wchevreuil @saintstack  Let me me if this looks good to you ?

**Comment by sandeepvinayak** [comment]
@saintstack  I believe this needs to be checked to master as well?

**Pull request merged**